presto-main/src/main/java/com/facebook/presto/execution/FutureStateChange.java

```diff
@@ -33,26 +33,14 @@
 @ThreadSafe
 public class FutureStateChange<T>
 {
-    // Use a separate future for each listener so canceled listeners can be removed
-    @GuardedBy("listeners")
-    private final Set<StateTrackingFuture<T>> listeners = new HashSet<>();
+    private final Listener listener = new Listener();
 
-    public ListenableFuture<T> createNewListener()
-    {
-        StateTrackingFuture<T> listener = StateTrackingFuture.create();
-        synchronized (listeners) {
-            listeners.add(listener);
-        }
-
-        // remove the listener when the future completes
-        listener.addListener(
-                () -> {
-                    synchronized (listeners) {
-                        listeners.remove(listener);
-                    }
-                },
-                directExecutor());
+    // Use a separate future for each listener so canceled futures can be removed
+    @GuardedBy("this")
+    private final Set<StateTrackingFuture<T>> futures = new HashSet<>();
 
+    public Listener getListener()
```

Comment:
Do you plan to introduce such mechanism in the future?

Reply:
I don't understand your question. But I think maybe the answer is in the second commit.

Reply:
Listener is a concrete class which is instantiated just once in FutureStateChange. Why not inline it into FutureStateChange?

Reply:
I want to hide `complete` and `fireStateChange`. This way, when people read code that uses the `Listener`, people can see that it's something you can use to subscribe to events.

Reply:
You could have `FutureStateChange` to be an interface that just exposes: `createNewListener`. Then this class would just be implementation (e.g: `CompletableStateChange`) with additional `complete` methods. Both approaches seems OK anyway.